zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+            + ", stopping the streaming.");
+      }
+    }
+
+    // Create output stream on the destination file and start streaming data
+    try {
+      outputStream = new FileOutputStream(destFile);
+
+      byte[] buffer = new byte[1024];
+      int lengthRead;
+      while ((lengthRead = inputStream.read(buffer)) > 0) {
+        outputStream.write(buffer, 0, lengthRead);
+        outputStream.flush();
+      }
+    } catch (FileNotFoundException e) {
+      throw new IOException(
+          "The destination local file " + destFile.getPath() + " could not be opened for write.");
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during data streaming while copying file " + backupFilePath
+              + " to local storage. The message of the IOException is: " + e.getMessage());
+    } finally {
+      inputStream.close();
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+  }
+
+  @Override
+  public void delete(File fileToDelete) throws IOException {
+    String backupFilePath = constructBackupFilePath(fileToDelete.getPath());
+    try {
+      if (!Files.deleteIfExists(Paths.get(backupFilePath))) {
+        throw new BackupException(
+            "File " + fileToDelete.getName() + "does not exist, cannot delete this file.");
+      }
+    } catch (DirectoryNotEmptyException e) {
+      throw new BackupException(
+          "Directory " + fileToDelete.getName() + "is not empty, cannot be deleted.");
+    }
+  }
+
+  /**
+   * Construct the file path in backup storage using file name
+   * @param fileName The file name for the file
+   * @return The file path to the location where the file will be stored in backup storage
+   */
+  private String constructBackupFilePath(String fileName) {
+    String backupFileName = fileName;
+
+    //Remove the temporary file name prefix in order to determine file type
+    if (fileName.startsWith(TMP_FILE_PREFIX)) {
+      backupFileName = fileName.replace(TMP_FILE_PREFIX, "");
+    }
+
+    // Get the directory name where the file would be stored based on the file type: snapshot or transaction log
+    String fileTypeDirName;
+    if (backupFileName.startsWith(Util.SNAP_PREFIX)) {
+      fileTypeDirName = Util.SNAP_PREFIX;
+    } else if (backupFileName.startsWith(Util.TXLOG_PREFIX)) {
+      fileTypeDirName = Util.TXLOG_PREFIX;
+    } else {
+      throw new IllegalArgumentException("No matching base file type for file " + fileName);
+    }
+    return String.join(PATH_SEPARATOR, _fileRootPath, fileTypeDirName, fileName);
+  }
+
+  /**
+   * Delete all the files whose file names starts with temporary file name prefix
+   * @param path The directory path for clean up
+   * @throws IOException
+   */
+  private void cleanupTempFiles(String path) throws IOException {
+    List<BackupFileInfo> tmpFiles = getBackupFileInfos(new File(path), TMP_FILE_PREFIX);
+    tmpFiles.forEach(fileInfo -> {
```

Comment:
Can we use stream() to parallel execute them?

Reply:
I think forEach() already parallel execute them. This is a description for this method from the Javadoc. Please correct me if I'm wrong.

> The behavior of this operation is explicitly nondeterministic. For parallel stream pipelines, this operation does not guarantee to respect the encounter order of the stream, as doing so would sacrifice the benefit of parallelism. For any given element, the action may be performed at whatever time and in whatever thread the library chooses. If the action accesses shared state, it is responsible for providing the required synchronization.